trapi_throttle/server.py

```diff
@@ -25,16 +39,154 @@
+
+    # Initialize the TAT
+    #
+    # TAT = Theoretical Arrival Time
+    # When the next request should be sent
+    # to adhere to the rate limit.
+    #
+    # This is an implementation of the GCRA algorithm
+    # More information can be found here:
+    # https://dev.to/astagi/rate-limiting-using-python-and-redis-58gk
+    await tat_db.set(
+        datetime.datetime.utcnow().isoformat()
+    )
+
+    # Use a new connection because subscribe method alters the connection
+    conn = await aioredis.create_redis(settings.redis_url, encoding = "utf-8")
+
+    # Subscribe to changes to the buffer
+    kp_buffer_pattern = f"__keyspace@0__:{kp_id}:buffer:*"
+    channel, = await conn.psubscribe(kp_buffer_pattern)
+
+    # TODO figure out why we need to sleep here
+    # for our tests to pass
+    await asyncio.sleep(1)
+
+    # Wait for anything to be added to the buffer
+    while await channel.wait_message():
+        await channel.get()
+
+        # Check if we actually have work to do
+        batch_keys = await APP.state.redis.keys(f"{kp_id}:buffer:*")
+        if len(batch_keys) == 0:
+            continue
+
+        LOGGER.debug(
+            f"Processing batch of size {len(batch_keys)} for KP {kp_id}"
+        )
+
+        now = datetime.datetime.utcnow()
+        tat = datetime.datetime.fromisoformat(await tat_db.get())
+
+        time_remaining_seconds = (tat - now).total_seconds()
+        # Wait for TAT
+        if time_remaining_seconds > 0:
+            LOGGER.debug(f"Waiting {time_remaining_seconds}")
+            await asyncio.sleep(time_remaining_seconds)
+
+        # Process batch
+        batch_request_ids = [key.split(':')[-1] for key in batch_keys]
+
+        request_values_db = [
+            RedisValue(APP.state.redis, f"{kp_id}:buffer:{request_id}")
+            for request_id in batch_request_ids
+        ]
+        request_values = await asyncio.gather(*[
+            request_value_db.get() for request_value_db in request_values_db
+        ])
+
+        # TODO Implement merging utility
+        merged_request_value = request_values[0]
+
+        async with httpx.AsyncClient() as client:
+            response = await client.post(kp_info.url, json = merged_request_value)
+
+        # TODO Implement split utility
+        for response_id in batch_request_ids:
+            await APP.state.redis.set(
+                f"{kp_id}:finished:{response_id}",
+                response.content,
+            )
+
+        # Update TAT
+        interval = kp_info.request_duration / kp_info.request_qty
+        new_tat = datetime.datetime.utcnow() + interval
+        await tat_db.set(new_tat.isoformat())
+
+    # Close redis connection
+    conn.close()
+    await conn.wait_closed()
+
+@APP.post("/register/{kp_id}")
+async def register_kp(
+        kp_id: str,
+        kp_info: KPInformation,
+        request: Request,
+):
+    kp_info_db = RedisValue(request.app.state.redis, f"{kp_id}:info")
```

Comment:
How is this different from using `APP.state.redis`?

Reply:
It's the same thing. I meant to switch to using `APP.state.redis` everywhere but I missed that one.